pkg/asset/agent/image/agentimage.go

```diff
@@ -30,56 +34,111 @@ func (a *AgentImage) Dependencies() []asset.Asset {
 	return []asset.Asset{
 		&Ignition{},
 		&BaseIso{},
+		&manifests.AgentManifests{},
+		&mirror.RegistriesConf{},
 	}
 }
 
 // Generate generates the image file for to ISO asset.
 func (a *AgentImage) Generate(dependencies asset.Parents) error {
 	ignition := &Ignition{}
-	dependencies.Get(ignition)
-
 	baseImage := &BaseIso{}
-	dependencies.Get(baseImage)
+	agentManifests := &manifests.AgentManifests{}
+	registriesConf := &mirror.RegistriesConf{}
+
+	dependencies.Get(ignition, baseImage, agentManifests, registriesConf)
 
 	ignitionByte, err := json.Marshal(ignition.Config)
 	if err != nil {
 		return err
 	}
 
-	ignitionContent := &isoeditor.IgnitionContent{Config: ignitionByte}
-	custom, err := isoeditor.NewRHCOSStreamReader(baseImage.File.Filename, ignitionContent, nil)
+	agentTuiFile, err := a.fetchAgentTuiBinary(agentManifests.ClusterImageSet.Spec.ReleaseImage, agentManifests.GetPullSecretData(), registriesConf.MirrorConfig)
+	if err != nil {
+		return err
+	}
+
+	err = a.prepareAgentISO(baseImage.File.Filename, ignitionByte, agentTuiFile)
 	if err != nil {
 		return err
 	}
 
-	a.imageReader = custom
 	a.cpuArch = ignition.CPUArch
 	a.rendezvousIP = ignition.RendezvousIP
 
 	return nil
 }
 
+func (a *AgentImage) fetchAgentTuiBinary(releaseImage string, pullSecret string, mirrorConfig []mirror.RegistriesConfig) (string, error) {
+	r := NewRelease(&executer.CommonExecuter{},
+		Config{MaxTries: OcDefaultTries, RetryDelay: OcDefaultRetryDelay},
+		releaseImage, pullSecret, mirrorConfig)
+
+	// TODO: This is just a temporary placeholder to test the entire workflow.
+	// As soon as https://github.com/openshift/assisted-installer-agent/pull/482 will land, then
+	// the following line should be fixed to extract the agent-tui binary
+	filename, err := r.ExtractFile("agent-installer-node-agent", "/usr/bin/agent")
+	if err != nil {
+		return "", err
+	}
+
+	return filename, nil
+}
+
+func (a *AgentImage) prepareAgentISO(iso string, ignition []byte, agentTuiFile string) error {
+	// Create a tmp folder to store all the pieces required to generate the agent ISO.
+	tmpPath, err := os.MkdirTemp("", "agent")
+	if err != nil {
+		return err
+	}
+	a.tmpPath = tmpPath
+
+	err = isoeditor.Extract(iso, a.tmpPath)
+	if err != nil {
+		return err
```

Comment:
I think it's a fair point. My initial approach was to remove always the temp folder to avoid filling up the space if the (failing) command is invoked by the user multiple times

Reply:
In general I do not expect to have failures very often here, we could eventually review this point later if we'll feel a more pressing need to have it